### PR TITLE
Add LineProcessProtocol and use it

### PR DESCRIPTION
**Pull request by p12tic**
This slightly simplifies the codebase by no longer needing to implement line handling in various places.

## Contributor Checklist:

* [x] I have updated the unit tests
* [not needed] I have created a file in the `master/buildbot/newsfragments` directory (and read the `README.txt` in that directory)
* [not needed] I have updated the appropriate documentation

### CODE REVIEW

File: master/buildbot/util/protocol.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+
+class LineBuffer:
+    def __init__(self, delimiter):
+        self._buffer = b''
+        self._delimiter = delimiter
+
+    def add_data(self, data):
+        # returns lines that have been processed, if any
+        lines = (self._buffer + data).split(self._delimiter)
```
Comment (tardyp) [inline master/buildbot/util/protocol.py]: you should rather use a generator for this. this is more scalable and pythonic

File: master/buildbot/util/protocol.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        self._buffer = lines.pop(-1)
+        return [l.rstrip(b'\r') for l in lines]
+
+    def get_trailing_lines(self):
```
Comment (tardyp) [inline master/buildbot/util/protocol.py]:
```suggestion
    def get_trailing_line(self):
```
there is only one line..

File: master/buildbot/util/protocol.py (hunk 1 above)
Comment (p12tic) [reply master/buildbot/util/protocol.py]: Fixed.

File: master/buildbot/util/protocol.py (hunk 2 above)
Comment (p12tic) [reply master/buildbot/util/protocol.py]: Fixed.

**Comment by p12tic** [comment]
@tardyp: This code is specific to Twisted ProcessProtocol. LogLineObserver has a little different requirements, so it does for example work only with decoded strings and does not handle lines that arrive split in the middle.

**Pull request merged**